### PR TITLE
Naren/multiphase src

### PR DESCRIPTION
**Pull request by nvohra0016**
Implemented general source term for multiphase flow; added additional 2D physical test for gas phase appearance due to external source term (multiphase_source.cc).

### CODE REVIEW

**Comment by lipnikov** [comment]
I suggest to rebase the code due to  too many failed commits.

File: src/pks/eos/IdealGas_Density.cc [hunk 1]
```diff
@@ -48,7 +48,8 @@ IdealGas_Density::DMolarDensityDp(double T, double p)
 void
 IdealGas_Density::InitializeFromPlist_()
 {
-  R_ = 8.31446261815324;
+  //R_ = 8.31446261815324;
```
Comment (lipnikov) [inline src/pks/eos/IdealGas_Density.cc:51]: I am confused about this change...

File: src/pks/multiphase/CMakeLists.txt [hunk 2]
```diff
@@ -91,7 +91,8 @@ set(multiphase_inc_files
   WRMmp_Corey.hh
   WRMmp_Simple.hh
   WRMmp_vanGenuchten.hh
-  WRMmp_BrooksCorey.hh)
+  WRMmp_BrooksCorey.hh
```
Comment (lipnikov) [inline src/pks/multiphase/CMakeLists.txt:94]: If this model is not-physical, I suggest to move it to the test directory, see flow/test/WRM_face.hh as an example. If it is physical, give it a better name.

File: src/pks/multiphase/EquationStructure.hh [hunk 3]
```diff
@@ -43,6 +43,9 @@ struct EquationStructure {
 
   Key storage;
   EquationTerm constraint;
+  
+  bool has_component = false;
+  int component_id; // Water has component_id = -1; other solutes have component_id \in \{0, 1, \dots}, sorted in the order as in "molecular diffusion"
```
Comment (lipnikov) [inline src/pks/multiphase/EquationStructure.hh:48]: I do not remember formatting rules for comments, it looks too long. Did you run (cd src; ../tools/formatting/clang-format.sh)?

File: src/pks/multiphase/Multiphase_PK.cc [hunk 4]
```diff
@@ -647,7 +645,7 @@ Multiphase_PK::Initialize()
   g_ = fabs(gravity_[dim_ - 1]);
 
   rho_l_ = S_->Get<double>("const_fluid_density");
-  eta_l_ = rho_l_ / CommonDefs::MOLAR_MASS_H2O;
+  eta_l_ = rho_l_ / mol_mass_H2O_;
```
Comment (lipnikov) [inline src/pks/multiphase/Multiphase_PK.cc:648]: Is it a place holder for future non-water based models?

File: src/pks/multiphase/Multiphase_TI.cc [hunk 5]
```diff
@@ -26,6 +26,7 @@
 #include "EquationStructure.hh"
 #include "Multiphase_PK.hh"
 #include "TotalComponentStorage.hh"
+#include "math.h"
```
Comment (lipnikov) [inline src/pks/multiphase/Multiphase_TI.cc:29]: Did you mean include <cmath> ?

File: src/pks/multiphase/Multiphase_TI.cc [hunk 6]
```diff
@@ -40,8 +41,17 @@ Multiphase_PK::FunctionalResidual(double t_old,
                                   Teuchos::RCP<TreeVector> u_new,
                                   Teuchos::RCP<TreeVector> f)
 {
-  double dtp = t_new - t_old;
+  double pi = M_PI;
```
Comment (lipnikov) [inline src/pks/multiphase/Multiphase_TI.cc:44]: I think this is not necessary. Use M_PI where needed.

File: src/pks/multiphase/WRMmp_Custom.cc [hunk 7]
```diff
@@ -0,0 +1,78 @@
+/*
```
Comment (lipnikov) [inline src/pks/multiphase/WRMmp_Custom.cc:1]: This is quite simple. I suggest to consider having only one .hh file for this model.

File: src/pks/multiphase/test/multiphase_jaffre.xml [hunk 8]
```diff
@@ -507,7 +543,7 @@
             <Parameter name="component" type="string" value="cell"/>
             <ParameterList name="function">
               <ParameterList name="function-constant">
-                <Parameter name="value" type="double" value="55408.51988"/>
+                <Parameter name="value" type="double" value="100000"/>
```
Comment (lipnikov) [inline src/pks/multiphase/test/multiphase_jaffre.xml:546]: Please double check this value.

File: src/pks/multiphase/test/multiphase_test.cc [hunk 9]
```diff
@@ -0,0 +1,151 @@
+/*
```
Comment (lipnikov) [inline src/pks/multiphase/test/multiphase_test.cc:1]: Please find a better name for this test.

**Review by lipnikov — COMMENTED** [review]
I suggest to rebase. See Fri Oct 20 commit as an example. The history of changes is preserved and there are no failed commits.

File: src/pks/eos/IdealGas_Density.cc (hunk 1 above)
Comment (nvohra0016) [reply src/pks/eos/IdealGas_Density.cc:51]: For manufactured solutions, I sometimes need to set R = 1 through the xml file. 
I have removed the commented line though.

File: src/pks/multiphase/Multiphase_PK.cc (hunk 4 above)
Comment (nvohra0016) [reply src/pks/multiphase/Multiphase_PK.cc:648]: Yes, and I also needed a custom defined value of molar mass of water for manufactured solution tests (specified through the xml file).

File: src/pks/multiphase/test/multiphase_jaffre.xml (hunk 8 above)
Comment (nvohra0016) [reply src/pks/multiphase/test/multiphase_jaffre.xml:546]: In [Gharbia, Jaffre' 14], for the MoMas benchmark test, the molar mass of water is 1e-2 kg/mol (as opposed to the usual 1.8e-2 kg/mol) and the mass density is 1e3 kg/m^3, which gives molar density 1e5 mol/m^3 (please see Table 1, Page 33 in the paper).

File: src/pks/multiphase/test/multiphase_test.cc (hunk 9 above)
Comment (nvohra0016) [reply src/pks/multiphase/test/multiphase_test.cc:1]: Removed; it was there by mistake.

File: src/pks/multiphase/Multiphase_TI.cc (hunk 6 above)
Comment (nvohra0016) [reply src/pks/multiphase/Multiphase_TI.cc:44]: Yes, removed.

File: src/pks/multiphase/Multiphase_TI.cc (hunk 5 above)
Comment (nvohra0016) [reply src/pks/multiphase/Multiphase_TI.cc:29]: Yes, thank you, removed.

File: src/pks/multiphase/CMakeLists.txt (hunk 2 above)
Comment (nvohra0016) [reply src/pks/multiphase/CMakeLists.txt]: Moved to test/, thank you.

File: src/pks/multiphase/WRMmp_Custom.cc (hunk 7 above)
Comment (nvohra0016) [reply src/pks/multiphase/WRMmp_Custom.cc]: Done, thank you.

File: src/pks/multiphase/EquationStructure.hh (hunk 3 above)
Comment (nvohra0016) [reply src/pks/multiphase/EquationStructure.hh]: I have reformatted the commented line.

**Comment by lipnikov** [comment]
Approved

**Review by lipnikov — APPROVED** [review]

**Pull request merged**